x/axelarnet/message_handler.go

```diff
@@ -0,0 +1,282 @@
+		return fmt.Errorf("chain %s is deactivated", destChain.Name)
+	}
+
+	if err := n.ValidateAddress(ctx, nexus.CrossChainAddress{Chain: destChain, Address: msg.DestinationAddress}); err != nil {
+		return err
+	}
+
+	switch msg.Type {
+	case nexus.TypeGeneralMessage:
+		return nil
+	case nexus.TypeGeneralMessageWithToken, nexus.TypeSendToken:
+		if !n.IsAssetRegistered(ctx, chain, token.GetDenom()) {
+			return fmt.Errorf("asset %s is not registered on chain %s", token.GetDenom(), destChain.Name)
+		}
+
+		if !n.IsAssetRegistered(ctx, destChain, token.GetDenom()) {
+			return fmt.Errorf("asset %s is not registered on chain %s", token.GetDenom(), destChain.Name)
+		}
+		return nil
+	default:
+		return fmt.Errorf("unrecognized message type")
+	}
+}
+
+// OnRecvMessage handles general message from a cosmos chain
+func OnRecvMessage(ctx sdk.Context, k keeper.Keeper, ibcK keeper.IBCKeeper, n types.Nexus, b types.BankKeeper, packet ibcexported.PacketI) ibcexported.Acknowledgement {
+	ack := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
+
+	data, err := toICS20Packet(packet)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	// skip if packet not sent to Axelar message sender account
+	if data.GetReceiver() != types.MessageSender.String() {
+		return ack
+	}
+
+	var msg Message
+	if err := json.Unmarshal([]byte(data.GetMemo()), &msg); err != nil {
+		ackErr := sdkerrors.Wrapf(types.ErrGeneralMessage, "cannot unmarshal memo")
+		return channeltypes.NewErrorAcknowledgement(ackErr)
+	}
+
+	// extract token from packet
+	token := extractTokenFromPacketData(packet)
+
+	path := types.NewIBCPath(packet.GetDestPort(), packet.GetDestChannel())
+
+	if err := validateMessage(ctx, k, n, path, msg, token); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	sourceChain := funcs.MustOk(k.GetChainNameByIBCPath(ctx, path))
+	sourceAddress := nexus.CrossChainAddress{
+		Chain:   funcs.MustOk(n.GetChain(ctx, sourceChain)),
+		Address: data.GetSender(),
+	}
+
+	switch msg.Type {
+	case nexus.TypeGeneralMessage:
+		err = handleMessage(ctx, n, sourceAddress, msg)
+	case nexus.TypeGeneralMessageWithToken:
+		err = handleMessageWithToken(ctx, n, b, ibcK, sourceAddress, msg, token)
+	case nexus.TypeSendToken:
+		err = handleTokenSent(ctx, n, b, ibcK, sourceAddress, msg, token)
+	default:
+		err = sdkerrors.Wrapf(types.ErrGeneralMessage, "unrecognized Message type")
+	}
+
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	return ack
+}
+
+func handleMessage(ctx sdk.Context, n types.Nexus, sourceAddress nexus.CrossChainAddress, msg Message) error {
+	destChain := funcs.MustOk(n.GetChain(ctx, nexus.ChainName(msg.DestinationChain)))
+
+	recipient := nexus.CrossChainAddress{Chain: destChain, Address: msg.DestinationAddress}
+	m := nexus.NewGeneralMessage(
+		n.GenerateMessageID(ctx, ctx.TxBytes()),
+		sourceAddress,
+		recipient,
+		crypto.Keccak256Hash(msg.Payload).Bytes(),
+		nexus.Sent,
```

Comment:
And the plan is for `ExecuteMessage` to set the status to `Sent` (as per the enum simplification), right? If destination is an EVM chain it'll add to the sent queue. If the destination is cosmos, it'd be similar, so we won't need a different type.